### PR TITLE
feat(git): Cache for local conflict detection

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

- Add repository-level caching for conflict check results

## Context:

- Ref: #3866
- Ref: #12575

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/git/index.ts [hunk 1]
```diff
@@ -625,6 +643,7 @@ export async function isBranchConflicted(
     }
   }
 
+  setCachedConflictResult(baseBranch, baseBranchSha, branch, branchSha, result);
```
Comment (viceice) [inline lib/util/git/index.ts:646]: Should we unset when we make a new commit ?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

**Review by viceice — DISMISSED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/git/index.ts [hunk 2]
```diff
@@ -585,14 +590,27 @@ export async function isBranchConflicted(
   logger.debug(`isBranchConflicted(${baseBranch}, ${branch})`);
   await syncGit();
   await writeGitAuthor();
-  if (!branchExists(baseBranch) || !branchExists(branch)) {
+
+  const baseBranchSha = getBranchCommit(baseBranch)?.slice(0, 7);
+  const branchSha = getBranchCommit(branch)?.slice(0, 7);
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/util/git/index.ts]: I think, it would be easier to just use long-form SHA, isn't it?

**Review by rarkins — APPROVED** [review]

File: lib/util/git/index.ts [hunk 3]
```diff
@@ -518,14 +523,27 @@ export async function isBranchConflicted(
...
   }
 
+  const cachedResult = getCachedConflictResult(
+    baseBranch,
+    baseBranchSha,
+    branch,
+    branchSha
+  );
+  if (is.boolean(cachedResult)) {
+    return cachedResult;
```
Comment (rarkins) [inline lib/util/git/index.ts:544]:
```suggestion
    logger.debug(`Using cached result ${cachedResult} for isBranchConflicted(${baseBranch}, ${branch})`);
    return cachedResult;
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/git/types.ts [hunk 4]
```diff
@@ -75,6 +75,22 @@ export interface CommitFilesConfig {
   force?: boolean;
 }
 
+type BranchName = string;
+type TargetBranchName = BranchName;
+type SourceBranchName = BranchName;
+
+export type GitConflictsCache = Record<TargetBranchName, TargetBranchConflicts>;
+
+interface TargetBranchConflicts {
```
Comment (viceice) [inline lib/util/git/types.ts:84]:
```suggestion
export interface TargetBranchConflicts {
```
We should export public interfaces when they are already exported implicit

File: lib/util/git/types.ts [hunk 5]
```diff
@@ -75,6 +75,22 @@ export interface CommitFilesConfig {
...
+  targetBranchSha: CommitSha;
+  sourceBranches: Record<SourceBranchName, SourceBranchConflict>;
+}
+
+interface SourceBranchConflict {
```
Comment (viceice) [inline lib/util/git/types.ts:89]:
```suggestion
export interface SourceBranchConflict {
```

File: lib/util/git/types.ts [hunk 6]
```diff
@@ -75,6 +75,22 @@ export interface CommitFilesConfig {
   force?: boolean;
 }
 
+type BranchName = string;
+type TargetBranchName = BranchName;
+type SourceBranchName = BranchName;
```
Comment (viceice) [inline lib/util/git/types.ts:80]:
```suggestion
export type BranchName = string;
export type TargetBranchName = BranchName;
export type SourceBranchName = BranchName;
```

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**